google-cloud-firestore/src/main/java/com/google/cloud/firestore/BulkWriter.java

```diff
@@ -568,28 +583,84 @@ public void run() {
    * next group of ready batches.
    */
   private void sendBatch(final BulkCommitBatch batch) {
-    boolean success = rateLimiter.tryMakeRequest(batch.getOperationCount());
+    boolean success = rateLimiter.tryMakeRequest(batch.getPendingOperationCount());
     Preconditions.checkState(success, "Batch should be under rate limit to be sent.");
-    try {
-      final ApiFuture<List<BatchWriteResult>> commitFuture = batch.bulkCommit();
-      commitFuture.addListener(
-          new Runnable() {
-            public void run() {
-              try {
-                batch.processResults(commitFuture.get(), null);
-              } catch (Exception e) {
-                batch.processResults(new ArrayList<BatchWriteResult>(), e);
+    MoreExecutors.directExecutor()
+        .execute(
+            new Runnable() {
+              public void run() {
+                bulkCommit(batch);
+                boolean removed = batchQueue.remove(batch);
+                Preconditions.checkState(removed, "The batch should be in the BatchQueue.");
+                sendReadyBatches();
               }
-              // Remove the batch from BatchQueue after it has been processed.
-              boolean removed = batchQueue.remove(batch);
-              Preconditions.checkState(removed, "The batch should be in the BatchQueue.");
-              sendReadyBatches();
-            }
-          },
-          MoreExecutors.directExecutor());
-    } catch (Exception e) {
-      batch.processResults(new ArrayList<BatchWriteResult>(), e);
+            });
+  }
+
+  private void bulkCommit(BulkCommitBatch batch) {
+    List<BatchWriteResult> results = new ArrayList<>();
+    for (int attempt = 0; attempt < MAX_RETRY_ATTEMPTS; ++attempt) {
+      final BulkCommitBatch finalBatch = batch;
+      ScheduledFuture<List<BatchWriteResult>> attemptBulkCommit =
+          firestore
+              .getClient()
+              .getExecutor()
+              .schedule(
+                  new Callable<List<BatchWriteResult>>() {
+                    public List<BatchWriteResult> call() {
+                      List<BatchWriteResult> results = new ArrayList<>();
+                      try {
+                        return finalBatch.bulkCommit().get();
+                      } catch (Exception e) {
+                        // Map the failure to each individual write's result.
+                        for (String path : finalBatch.getPendingDocs()) {
+                          if (e instanceof FirestoreException) {
+                            results.add(
+                                new BatchWriteResult(
+                                    path,
+                                    null,
+                                    ((FirestoreException) e).getStatus(),
+                                    e.getMessage()));
+                          } else {
```

Comment:
I wasn't sure how to handle the case where a non-FirestoreException was thrown. I added an additional field for an Exception in the event that the Exception thrown does not have a `Status`, but I wonder if this could be cleaner.

Reply:
The code in `bulkCommit` should wrap all exceptions in FirestoreException.